### PR TITLE
cni: revisioned install-cni & istio-cni binaries

### PR DESCRIPTION
**Pull request by eoinfennessy**
- Enables multiple install-cni daemonsets and istio-cni plugins to be installed & run on each node
- Appends the Istio revision to the install-cni daemonset, istio-cni binaries, and cni config
- Each plugin instance checks the revision of the pod it is being invoked against and skips setting up the redirect if the revision does not match its own

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

**Review by bleggett — CHANGES_REQUESTED** [review]
This PR won't work with ambient and would stop working entirely once https://github.com/istio/istio/pull/48253 is merged.

Though I believe we could fix that, the approach itself is not unsound. There's nothing strictly wrong with installing multiple plugins keyed by revision on the node so I'm not against the approach conceptually - right now the only way to update the CNI is via node cordon which is foolproof and simple but inflexible.

@eoinfennessy Can you talk thru this during [one of our WG meetings](https://docs.google.com/document/d/1wsa06GGiq1LEGwhkiPP0FKIZJqdAiue-VeBonWAzAyk/)? I want to understand what operational problem you're trying to solve, exactly. It's an interesting proposal but others may have thoughts.

- We had previously decided that node cordons are the recommended way to manage CNI upgrades (I'm open to changing that stance but it would be good to make the argument as to why node cordons are insufficient for your use case).

- If we do something like this, it also must work for ambient, as the CNI controller is shared between ambient and sidecar.

- If we do something like this, we must account for the fact that the CNI controller on the node interacts with other node-level things (CNI config, iptables rules in the host netns) that are not currently revisioned - it's more than just the plugin.

- There are as-yet-unsolved issues around adding/removing CNI plugins on nodes - more agent instances adding and removing more plugins exacerbates that, e.g. https://github.com/istio/istio/issues/46848

- CNI is not in the datapath and has no controlplane dep, so it should be possible to upgrade CNI in-place without affecting workloads (inpod actually makes this even simpler for ambient, since host-level redirection is out of the picture) - which IMO lessens the need to support a SxS/revisioned CNI agent proper.

File: manifests/charts/istio-cni/templates/daemonset.yaml
Comment (yuval-k) [inline manifests/charts/istio-cni/templates/daemonset.yaml:31]: usually revision is added as a separate label. see the label in the daemonset (line 17 in this file)

**Review by howardjohn — CHANGES_REQUESTED** [review]
One way you could do this today, which I would generally recommend is to run separate node pools per revision and isolate that way.

In general, this needs to be brought up as a design doc with use cases, requirements, and motivations before we proceed along any implementation.

**Comment by eoinfennessy** [comment]
@bleggett, thanks for the info regarding ambient compatibility and work on [#48253](https://github.com/istio/istio/pull/48253). It makes sense to wait until that gets merged before doing any further work on this implementation. In the meantime, I'll prepare a doc to share & discuss with the Istio community outlining requirements, motivations, and use cases.

In short, this work aims to help support canary-based upgrades without requiring node cordoning/node pools, and may have use cases in soft multi-tenancy environments where tenants require independent upgrades to Istio and CNI.

If my understanding is correct, after #48253 merges, ambient mode will no longer configure iptables rules on the host netns. Perhaps this could help to make revision support for CNI possible in ambient mode? Having said that, the possibility of scoping this work to non-ambient use cases only could also be considered.

File: manifests/charts/istio-cni/templates/daemonset.yaml
Comment (eoinfennessy) [reply manifests/charts/istio-cni/templates/daemonset.yaml:31]: The revision is appended here to provide unique `selector.matchLabels` for pods created by each install-cni DS instance.

**Comment by howardjohn** [comment]
> @howardjohn, the basic use case is for supporting canary rollouts where there may have been changes to the cni layer. It's hard for me to see how cordoning off nodes would help, as user applications would also need to be cordoned to the same nodes running the updated cni. Maybe I'm missing something?

Yes, that is exactly how it would work.

Do we really have users that want to run multiple daemonsets on every single node for every revision? Soon we will fill up the whole node just to `cp` a binary onto the host...

What changes are they worried about in CNI? Is it theoretical issues or legitimate ones? CNI has been incredibly stable with minimal changes for years

**Comment by rcernich** [comment]
It's certainly theoretical, but I don't know the risk in practice, nor the impact should changes need to be made.  I suppose, if changes are required, then you could force something like this to go along with those changes at that time.

The daemon sets should be removed once the older version of the control plane is torn down, so I don't think you'll have a bunch of istio-cni revs running forever.  That would just be rude :)

**Comment by rcernich** [comment]
To add some background, the new operator we are working on will support automation of canary rollouts of the control plane (creating a new control plane and deleting the old control plane after it is no longer managing any proxies, with a following grace period; users will still need to manually migrate their namespaces and applications to the new control plane).  As part of this work, we are going through and addressing issues we are finding in the existing revisioning.  CNI is one of those components that currently does not support revisioning, hence this PR.  HTH.

**Comment by linsun** [comment]
> To add some background, the new operator we are working on will support automation of canary rollouts of the control plane (creating a new control plane and deleting the old control plane after it is no longer managing any proxies, with a following grace period; users will still need to manually migrate their namespaces and applications to the new control plane). As part of this work, we are going through and addressing issues we are finding in the existing revisioning. CNI is one of those components that currently does not support revisioning, hence this PR. HTH.

With sidecars (no ambient), do you have users requiring revisions of istio-cni?  We don't see any users asking for it.  In place upgrade for istio-cni seems working well for our users.

**Review by linsun — COMMENTED** [review]
My preference is to introduce revision of istio-cni only when we absolutely need it and we had this component for 4-5 years in Istio for sidecars. I think users love in place for istio-cni and would rather stay with in place upgrade for simplicity.

**Comment by rcernich** [comment]
@linsun , in the past we have had to fix bugs in cni well after ..0 releases (as we tend to lag on the product side).  This leads me to believe that cni is not used broadly across the community, or users of cni also lag, or they patch them independently.

Regarding in-place upgrades, we are getting more and more requests to support canary rollouts.  Being able to support that through an operator makes the process of managing them much simpler.

Lastly, as revisions are a supported feature, I believe that ensuring that they are properly implemented across the project should be a priority.  It doesn't do any good to add support _after_ you need it.

I'd also like to hear why people think this is a risky change.  To me, this shouldn't be very controversial.  It's adding support that should have been there when revisions were first introduced.  I'd say it's riskier to not have it and we've been lucky up to this point.

**Comment by howardjohn** [comment]
> @linsun , in the past we have had to fix bugs in cni well after ..0 releases (as we tend to lag on the product side). This leads me to believe that cni is not used broadly across the community, or users of cni also lag, or they patch them independently.

FWIW our product (more or less) mandates CNI usage

> Lastly, as revisions are a supported feature, I believe that ensuring that they are properly implemented across the project should be a priority. It doesn't do any good to add support after you need it.

It _is_ correctly implemented per the projects (current) view of "correct". Lack of revisions in CNI is not an oversight but an intentional design decision. CNI is quite different from Istiod and the same principals that apply to Istio do not necessarily apply to CNI

**Comment by rcernich** [comment]
@howardjohn , is that documented somewhere, e.g. in an RFC?  I recall having many conversations in the environments working group about how global resources should be handled with revisions (e.g. What do we do if we add/remove permissions from a ClusterRole?), and the feature was really rolled out over several releases as these types of problems were encountered and resolved.

As far as CNI goes, any changes to ports, etc., on the proxy will necessitate a change to CNI, and this has happened in the past.  If we want to restrict those types of changes and mark those settings as API, then...

**Comment by bleggett** [comment]
> Regarding in-place upgrades, we are getting more and more requests to support canary rollouts. Being able to support that through an operator makes the process of managing them much simpler.

Yep, and I share your desire to support canary rollouts. Canary rollouts exist to prevent disruption to existing user workloads during an upgrade. The goal with ambient at least, and the recent inpod work, is to disconnect the CNI agent (and thus plugin) entirely from the datapath of existing user workloads, so that the CNI agent can be upgraded in-place without disruption to existing workloads. It is already disconnected from the control plane.

In general I prefer revisions as a way to manage canary deployments of Istio and give users explicit blast radius control, but revisions are for blast radius control _existing_ workload traffic. Which means things like the control plane and the data plane, arguably - but not the CNI, if we are doing things correctly.

> As far as CNI goes, any changes to ports, etc., on the proxy will necessitate a change to CNI, and this has happened in the past. If we want to restrict those types of changes and mark those settings as API, then...

https://istio.io/latest/docs/setup/additional-setup/cni/#upgrade

Here we say 

> The CNI plugin at version 1.x is compatible with control plane at version 1.x-1, 1.x, and 1.x+1, which means CNI and control plane can be upgraded in any order, as long as their version difference is within one minor version.

IMO that should read "The CNI **agent**" but either way that's the current standard for the CNI-CP skew we sign up to support.

**Comment by eoinfennessy** [comment]
Closing. After discussions regarding the Sail operator & CNI at yesterday's WG meeting, Red Hat have decided that a global CNI config will be offered instead of this revision-based approach.

**Comment by linsun** [comment]
Thank you @eoinfennessy for your understanding - pls keep us posted on how this goes.

**Comment by bleggett** [comment]
Also opened https://github.com/istio/istio/issues/49009 to track this more broadly

**Pull request closed without merging**
